### DISCUSSION TITLE
Hitting "pull" key twice showing modal popup is a bit annoying

### DISCUSSION THREAD

**Discussion by peppy** [jesseduffield/lazygit#3116] [General]
![CleanShot 2023-11-17 at 03 13 26](https://github.com/jesseduffield/lazygit/assets/191335/5661ddd4-c25f-48ee-b72e-0739b77c479b)

I do this more often than I'd like to admit to. It requires explicit dismissal and can sometimes interrupt flow (ie. if I've already pressed keys for my next actions without realising i pulled twice).

**Comment by stefanhaller** [comment]
I can see how this can be annoying when you type fast, but I'm not sure there's much we can do about it; we do want to show an error message if you type the key for a disabled command.

Maybe we could consider showing the error in a toast in the status bar instead of as a model panel. @jesseduffield What do you think?

**Reply by jesseduffield** [reply]
Introducing an error variant of our toast is an interesting idea: plenty of other apps use error toasts and it's great to not have to explicitly acknowledge them. It would be trivially easy to implement i.e. just colour it red. So I'm in favour.

_But_ we would need to be very clear on the criteria that decides whether an error should be a toast or a popup, so that contributors know which one to use in a given situation. Maybe:
* would the user be annoyed from having to explicitly acknowledge the error?
* is it fine if the user doesn't notice the error?

Those criteria are satisfied for the case of double pushing/pulling.

Thoughts?

**Reply by stefanhaller** [reply]
Initially I'd use the toast for the general case that the user is trying to invoke a disabled command. We don't get to decide this on a case-by-case basis, it would be for all of them. Later we could decide to use toasts instead of popups for other cases as well, and only then would we need clear criteria.

BTW, I'm not sure I agree with the "trivially easy to implement"; we'd also need to call `utils.Decolorise` on the string during layout, and because of that I wouldn't want to work on it before #3076 is merged, because it would conflict with that one.

**Reply by jesseduffield** [reply]
That makes sense

**Reply by stefanhaller** [reply]
There's a draft implementation of this at #3180.
